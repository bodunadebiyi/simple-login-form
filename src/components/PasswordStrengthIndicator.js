import React from 'react';
import injectSheet from 'react-jss';
import PropTypes from 'prop-types';

const styles = {
  wrapper: {
    width: '100%',
    textAlign: 'left',
    maxWidth: 340,
    display: 'flex',
    flexDirection: 'column',
  },
  bar: {
    height: 5,
    width: '30%',
    maxWidth: 100,
  },
  bars: {
    display: 'flex',
    flexDirection: 'row',
    justifyContent: 'space-between',
    '& div': {
      backgroundColor: '#ccc',
    }
  },
  passwordIsWeak: {
    '& > div:first-child': {
      backgroundColor: 'red',
    }
  },
  passwordIsGood: {
    '& > div:nth-child(-n+2)': {
      backgroundColor: 'orange',
    }
  },
  passwordIsVeryGood: {
    '& div': {
      backgroundColor: 'green',
    }
  },
  passwordVerdict: {
    fontWeight: 'bold',
    fontSize: 12,
    color: 'black',
    marginTop: 5
  }
}

const passwordStrength = {
  INVALID: 0,
  WEAK: 1,
  GOOD: 2,
  VERY_GOOD: 3
};

const PasswordStrengthIndicator = React.memo(({
  classes,
  strength
}) => {
  let barStyle = `${classes.bars} `;
  let passwordVerdict;

  switch(strength) {
    case passwordStrength.INVALID:
      passwordVerdict = 'Invalid';
      break;
    case passwordStrength.WEAK:
      passwordVerdict = 'Weak';
      barStyle += classes.passwordIsWeak;
      break;
    case passwordStrength.GOOD:
      passwordVerdict = 'Good';
      barStyle += classes.passwordIsGood;
      break;
    case passwordStrength.VERY_GOOD:
      passwordVerdict = 'Very Good';
      barStyle += classes.passwordIsVeryGood;
      break;
    default:
      passwordVerdict = '';
      barStyle += classes.invalidBar;
  }

  return (
    <div className={classes.wrapper}>
      <div className={barStyle} id='bars'>
        <div className={classes.bar} />
        <div className={classes.bar} />
        <div className={classes.bar} />
      </div>
      <span className={classes.passwordVerdict}>
        Strength: {passwordVerdict}
      </span>
    </div>
  )
});

PasswordStrengthIndicator.defaultProps = {
  strength: 3
};

PasswordStrengthIndicator.props = {
  strength: PropTypes.number.isRequired,
}

export default injectSheet(styles)(PasswordStrengthIndicator);
