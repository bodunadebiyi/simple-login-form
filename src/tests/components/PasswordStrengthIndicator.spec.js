import React from 'react';
import { shallow } from 'enzyme';
import PasswordStrengthIndicator from '../../components/PasswordStrengthIndicator';

describe('PasswordStrengthIndicator', () => {
  describe('when component is mounted', () => {
    it('renders without crashing', () => {
      const component = shallow(<PasswordStrengthIndicator />);
      expect(component.exists()).toBe(true);
    });

    it('matches snapshot', () => {
      const component = shallow(<PasswordStrengthIndicator strength={0} />);
      expect(component).toMatchSnapshot();
    })
  });

  describe('when password strength is 0', () => {
    it('displays a password verdict of "Invalid"', () => {
      const component = shallow(<PasswordStrengthIndicator strength={0} />);
      const passwordVerdictText = component.dive().find('span').props().children;

      expect(passwordVerdictText.includes('Invalid')).toBe(true);
    });
  });

  describe('when password strength is 1', () => {
    let component;
    
    beforeEach(() => {
      component = shallow(<PasswordStrengthIndicator strength={1} />);
    });

    it('displays a password verdict of "Weak"', () => {
      const passwordVerdictText = component.dive().find('span').props().children;
      expect(passwordVerdictText.includes('Weak')).toBe(true);
    });

    it('expects bars div to contain a className of passwordIsWeak', () => {
      const barsDivClassName = component.dive().find('#bars').props().className;
      expect(barsDivClassName.includes('passwordIsWeak')).toBe(true);
    })
  });

  describe('when password strength is 2', () => {
    let component;
    
    beforeEach(() => {
      component = shallow(<PasswordStrengthIndicator strength={2} />);
    });

    it('displays a password verdict of "Good"', () => {
      const passwordVerdictText = component.dive().find('span').props().children;
      expect(passwordVerdictText.includes('Good')).toBe(true);
    });

    it('expects bars div to contain a className of passwordIsGood', () => {
      const barsDivClassName = component.dive().find('#bars').props().className;
      expect(barsDivClassName.includes('passwordIsGood')).toBe(true);
    })
  });

  describe('when password strength is 3', () => {
    let component;
    
    beforeEach(() => {
      component = shallow(<PasswordStrengthIndicator strength={3} />);
    });

    it('displays a password verdict of "VeryGood"', () => {
      const passwordVerdictText = component.dive().find('span').props().children;
      expect(passwordVerdictText.includes('Very Good')).toBe(true);
    });

    it('expects bars div to contain a className of passwordIsVeryGood', () => {
      const barsDivClassName = component.dive().find('#bars').props().className;
      expect(barsDivClassName.includes('passwordIsVeryGood')).toBe(true);
    })
  });
});