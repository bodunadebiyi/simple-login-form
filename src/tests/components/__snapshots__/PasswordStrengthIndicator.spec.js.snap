// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PasswordStrengthIndicator when component is mounted matches snapshot 1`] = `ShallowWrapper {}`;
